test/fenv/fenv_gentests.cc

```diff
@@ -0,0 +1,315 @@
+  out << "    failures += !(result_passed && exceptions_passed);\n";
+  out << "    fesetenv( &fenv );\n";
+  out << "  },\n";
+  std::fesetenv( &fenv );
+
+  if( ++testcount >= maxtests_per_file ) {
+    closefile();
+    testcount = 0;
+  }
+}
+
+template<typename FP>
+static const FP special_fp_values[] = {
+  0.0f,
+  -0.0f,
+  1.0f,
+  -1.0f,
+  1.5f,
+  -1.5f,
+  std::numeric_limits<FP>::quiet_NaN(),
+  std::numeric_limits<FP>::signaling_NaN(),
+  std::numeric_limits<FP>::infinity(),
+  -std::numeric_limits<FP>::infinity(),
+  std::numeric_limits<FP>::lowest(),
+  std::numeric_limits<FP>::max(),
+};
+
+template<typename FP, typename INT>
+static const FP special_fcvt_values[] = {
+  FP( std::numeric_limits<INT>::max() ),
+  FP( std::numeric_limits<INT>::max() ) + FP( 0.5 ),
+  FP( std::numeric_limits<INT>::max() ) + FP( 1.0 ),
+  FP( std::numeric_limits<INT>::max() ) - FP( 0.5 ),
+  FP( std::numeric_limits<INT>::max() ) - FP( 1.0 ),
+  FP( std::numeric_limits<INT>::min() ),
+  FP( std::numeric_limits<INT>::min() ) + FP( 0.5 ),
+  FP( std::numeric_limits<INT>::min() ) + FP( 1.0 ),
+  FP( std::numeric_limits<INT>::min() ) - FP( 0.5 ),
+  FP( std::numeric_limits<INT>::min() ) - FP( 1.0 ),
+  FP( 0 ),
+  -FP( 0 ),
+  FP( 0.1 ),
+  FP( -0.1 ),
+  FP( 0.5 ),
+  FP( -0.5 ),
+  FP( 0.9 ),
+  FP( -0.9 ),
+  std::numeric_limits<FP>::quiet_NaN(),
+  std::numeric_limits<FP>::signaling_NaN(),
+  std::numeric_limits<FP>::infinity(),
+  -std::numeric_limits<FP>::infinity(),
+};
+
+#define OPER_PAIR( lambda ) \
+  { lambda, #lambda }
+
+template<typename FP, typename INT>
+const char fcvt_instruction[] = "";
+template<>
+const char fcvt_instruction<float, int32_t>[] = "fcvt.w.s";
+template<>
+const char fcvt_instruction<float, uint32_t>[] = "fcvt.wu.s";
+template<>
+const char fcvt_instruction<double, int32_t>[] = "fcvt.w.d";
+template<>
+const char fcvt_instruction<double, uint32_t>[] = "fcvt.wu.d";
+
+#if __riscv_xlen >= 64
+template<>
+const char fcvt_instruction<float, int64_t>[] = "fcvt.l.s";
+template<>
+const char fcvt_instruction<float, uint64_t>[] = "fcvt.lu.s";
+template<>
+const char fcvt_instruction<double, int64_t>[] = "fcvt.l.d";
+template<>
+const char fcvt_instruction<double, uint64_t>[] = "fcvt.lu.d";
+#endif
+
+template<typename FP, typename INT>
+void generate_fcvt_tests() {
+  char test_src[256];
+  test_src[0] = 0;
+  strcat( test_src, "[]( volatile auto x ) { " );
+  strcat( test_src, type<INT> );
+  strcat( test_src, " res; asm volatile( \"" );
+  strcat( test_src, fcvt_instruction<FP, INT> );
+  strcat( test_src, R"( %0, %1" : "=r"(res) : "f"(x) ); return res; })" );
+
+  using INT_FUNC1 = std::pair<INT ( * )( FP ), const char*>[];
+  for( auto oper_pair : INT_FUNC1{
+         {[]( volatile auto x ) { return to_int<INT>( x ); }, test_src},
+  } ) {
+    for( auto x : special_fcvt_values<FP, INT> ) {
+      generate_test( oper_pair, x );
+    }
+  }
```

Comment:
For FP -> Integer conversions, we use the `fcvt` instruction corresponding to the source and destination types, and call it in inline assembly.